packages/massa-web3/src/experimental/smartContract.ts

```
@@ -225,12 +239,138 @@ export class SmartContract {
+   */
+  async ensureBalance(
+    operationName: string,
+    account: Account,
+    coins?: bigint
+  ): Promise<void> {
+    if (coins) {
+      const balance = await this.client.getBalance(account.address.toString())
+      if (balance < coins) {
+        throw new InsufficientBalanceError({
+          operationName,
+          userBalance: balance,
+          neededBalance: coins,
+        })
+      }
+    }
+  }
+
+  /**
+   * Return the max gas amount for an operation, ensuring it is within the allowable range.
+   *
+   * @param operationName - The name of the operation where the gas usage is being checked.
+   * @param maxGas - The initial max gas amount proposed for the operation.
+   * @param min - The minimum allowable gas limit for the operation.
+   * @param max - The maximum allowable gas limit for the operation.
+   *
+   * @returns The validated gas amount.
+   *
+   * @throws MaxGasError if the proposed gas amount is either too high or too low.
+   */
+  async getMaxGas(
+    operationName: string,
+    maxGas: bigint,
+    min: bigint,
+    max: bigint
+  ): Promise<bigint> {
+    if (!maxGas) return await this.getGasEstimation()
+
+    if (maxGas > max) {
+      throw new MaxGasError({ operationName, isHigher: true, amount: max })
+    } else if (maxGas < min) {
+      throw new MaxGasError({ operationName, isHigher: false, amount: min })
+    }
+    return maxGas
+  }
+
+  /**
+   * Returns the fee for an operation, ensuring it is at least the minimal fee required.
+   *
+   * @param operationName - The name of the operation for which the fee is being checked. This helps identify the context in errors.
+   * @param fee - The fee proposed for the operation. If no fee is provided, the minimal fee will be used.
+   *
+   * @returns The validated or minimal fee.
+   *
+   * @throws MinimalFeeError if the fee provided is less than the minimum required fee.
+   */
+  async getFee(operationName: string, fee: bigint): Promise<bigint> {
```

not a fan of passing the name on parameters just for the error message. you should throw a generic error here, and eventually catch it where its needed to add the operation name in the error message

---

We should have one single way of handling fee for every operation. Make the changes for that Transfers; buy roll/ sell roll etc use this implementation